superset/datasource/dao.py

```diff
@@ -0,0 +1,59 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+from typing import Dict, Type, Union
+
+from sqlalchemy.orm import Session
+
+from superset.connectors.sqla.models import SqlaTable
+from superset.dao.base import BaseDAO
+from superset.dao.exceptions import DatasourceNotFound, DatasourceTypeNotSupportedError
+from superset.datasets.models import Dataset
+from superset.models.sql_lab import Query, SavedQuery
+from superset.tables.models import Table
+from superset.utils.core import DatasourceType
+
+Datasource = Union[Dataset, SqlaTable, Table, Query, SavedQuery]
+
+
+class DatasourceDAO(BaseDAO):
+
+    sources: Dict[DatasourceType, Type[Datasource]] = {
+        DatasourceType.TABLE: SqlaTable,
+        DatasourceType.QUERY: Query,
+        DatasourceType.SAVEDQUERY: SavedQuery,
+        DatasourceType.DATASET: Dataset,
+        DatasourceType.SLTABLE: Table,
+    }
+
+    @classmethod
+    def get_datasource(
+        cls, session: Session, datasource_type: DatasourceType, datasource_id: int
+    ) -> Datasource:
+        if datasource_type not in cls.sources:
```

Comment:
I think we're going to want to accept the string value as well here. 
I don't think `"table" in cls.sources` is going to work for example.

Reply:
I'm currently casting all the values going into the `get_datasource` function like this:
```
DatasourceDAO.get_datasource(
  db.session, DatasourceType(datasource["type"]), int(datasource["id"])
)
```
So moving forward i think we should push for the enums since it's easy to cast via just wrapping it in a `DatasourceType()` constructor. If we allow string we'd have to add logic to handle/report whenever the user sending us invalid values.

Reply:
you could also just wrap `source = cls.sources[datasource_type]` in a try/catch block. I just think there's too much room for error here across the application with people who may not know that they can't pass in a string.

Reply:
you could also just wrap `source = cls.sources[datasource_type]` in a try/catch block. I just think there's too much room for error here across the application with people who may not know that they can't pass in a string. This should work if datasource_type is either a string or the enum key.